### PR TITLE
Cascading Policies

### PR DESCRIPTION
**Pull request by josephlewis42**
This PR adds a generic "rules engine" to the broker. Rather than being a full rules-engine, it behaves more like CSS. Rules are executed top to bottom and overwrite properties as they match.

The rules list also supports user-defined assertions which are executed when the rules are unmarshalled from their source. This lets the users enter rudimentary test cases that will be validated by the service on startup.

There are three immediate use-cases for this engine:

* Provision overrides
* Bind overrides
* Environment variable overrides (to add multi-project provisioning support)

Due to the service being extensible, provision and bind overrides can no longer be handled in a generic way through the forms.

User entered rules will look something like this for a fizzbuzz based example:

```json
{
  "policy": [
    {
      "//": "by default print the number",
      "then": {"print": "{{number}}"}
    },
    {
      "if": {"multiple-of-3": "true"},
      "then": {"print": "fizz"}
    },
    {
      "if": {"multiple-of-5": "true"},
      "then": {"print": "buzz"}
    },
    {
      "//": "multiples of 3 and 5 print fizzbuzz",
      "if": {"multiple-of-3": "true", "multiple-of-5": "true"},
      "then": {"print": "fizzbuzz"}
    }
  ],
  "assert": [
    {
      "//": "multiple of 3 and 5",
      "if": {"multiple-of-3": "true", "multiple-of-5": "true"},
      "then": {"print": "fizzbuzz"}
    },
    {
      "//": "not multiple of 3 or 5",
      "if": {"multiple-of-3": "false", "multiple-of-5": "false"},
      "then": {"print": "{{number}}"}
    },
  ]
}
```

### CODE REVIEW

File: pkg/broker/policy/policy.go [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+	allowedSet := utils.NewStringSet(allowedKeys...)
+	condKeys := utils.NewStringSetFromStringMapKeys(cond)
+
+	invalidKeys := condKeys.Minus(allowedSet)
+
+	if invalidKeys.IsEmpty() {
+		return nil
+	}
+
+	return fmt.Errorf("unknown condition keys: %v condition keys must one of: %v, check their capitalization and spelling", invalidKeys, allowedKeys)
```
Comment (craigatgoogle) [inline pkg/broker/policy/policy.go]: s/must one of/must be one of

**Comment by josephlewis42** [comment]
@craigatgoogle sure thing! The plan is to enable this so users can override Brokerpak environment variables so they can do things like target different projects with different service/space/organization combinations:

```
{
  "policy": [
    {
      "name": "by default the project is default-project",
      "effect": {"GCP_PROJECT_ID": "default-project"}
    },
    {
      "condition": {"service_name": "google-cloudsql-mysql"},
      "effect": {"GCP_PROJECT_ID": "dba-owned-project"}
    },
    {
      "//": "Developer sandbox space",
      "if": {"space_id": "62db8b06-134a-4fb1-958d-74b7fce12222"},
      "then": {
        "GCP_PROJECT_ID": "developer-free-for-all-project",
        "GCP_SERVICE_ACCOUNT":"{...}"
      }
    }
  ],
  "assert": [
    {
      "//": "developer provisioning into sandbox",
      "if": {"space_id": "62db8b06-134a-4fb1-958d-74b7fce12222", "service_name": "google-cloudsql-mysql"},
      "then": {"GCP_PROJECT_ID": "developer-free-for-all-project", "GCP_SERVICE_ACCOUNT":"{...}"}
    }
  ]
}
```

File: pkg/broker/policy/policy_test.go [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+		},
+		"good-fizzbuzz": {
+			Json: `{
+			"policy": [
+				{"if": {}, "then": {"print":"{{number}}"}},
+				{"if": {"multiple-of-3":"true"}, "then": {"print":"fizz"}},
+				{"if": {"multiple-of-5":"true"}, "then": {"print":"buzz"}},
+				{"if": {"multiple-of-3":"true", "multiple-of-5":"true"}, "then": {"print":"fizzbuzz"}}
+			],
+			"assert": [{"if":{"multiple-of-3":"true", "multiple-of-5":"true"}, "then":{"print":"fizzbuzz"}}]
```
[the remark this answers is not in the document]
Comment (josephlewis42) [reply pkg/broker/policy/policy_test.go]: There are some examples up above in `TestPolicyList_Apply` where values cascade together. 

A real-world example would be setting a JSON credential for a particular organization and each space would get a distinct project id.

**Comment by craigatgoogle** [comment]
Thanks, that does provide some further clarity.

**Pull request merged**